### PR TITLE
Add castable types

### PR DESCRIPTION
**Pull request by xujustinj**
Adds an interface to cast data from one type to another via Pydantic, but doesn't use it for anything just yet.
Nevertheless this is fully tested with the power of Cursor.

## Why??

In C++, you can take any variable `x` and pass it into any function parameter so long as there is an implemented type conversion from the type of `x` to the type expected by the parameter.
This is super powerful and can even happen recursively in generic types like C++ vectors.

We want to enable this with our nodes, but Python isn't exactly C++ so we can't just do a compile time conversion.
Instead, we'll need a unified way to send things from one type to another.
Thus, we introduce the `Value` class, hacked together from Pydantic's `RootModel`.

Each node will take, as input, a string-`Value` object, and it will also output a string-`Value` object.
When we pass a value from the output of one node to the input of another node, we will attempt to typecast it over.
This will remove the need to precisely control a node's output format, since the user will just connect whatever they want as long as an obvious typecast exists.

## What abilities does type casting give me?

```py
five_str = StringValue(root="5")
five_int = five_str.cast_to(IntegerType, context=context)  # IntegerType(root=5)
# or
# five_int = IntegerType.cast_from(five_str, context=context)
```

That looks cute but kind of silly since you can also just call `int("5")` to get `5`.

Here's an ability that some might consider unnatural: we can use a _context_ to provide extremely powerful type casts.
Coming in a later PR, we might have something like:

```py
files = ArrayValue[PDFFileValue].model_validate_json("""
  [
    {"path": "somewhere/in/our/supabase/bucket"},
    {"path": "somewhere/else/in/our/supabase/bucket"}
  ]
""")
text = file.cast_to(ArrayValue[StringValue], context=our_supabase_context)  # reads every PDF using docling
```

## What does it need in return?

We enable casting via the `add_cast` method.

```py
SourceValue.add_cast(TargetValue, lambda S, T: lambda v, ctx: TargetValue(root=...))
```

`S` and `T` are advanced arguments that only really apply for generic types, where the type casting function depends on the element type of a generic type.
For the most part, you can ignore them, but if you must (e.g. if you're reviewing this PR), check out the `ArrayValue` or `StringMapValue` types.

### CODE REVIEW

File: src/workflow_engine/core/data.py [hunk 1]
```diff
@@ -8,15 +8,13 @@ class Data(BaseModel):
...
 
-Output = TypeVar("Output", bound=Data)
 Output_co = TypeVar("Output_co", bound=Data, covariant=True)
 
 
 __all__ = [
     "Data",
-    "Input",
-    "Output",
+    "Input_contra",
```
Comment (sunapi386) [inline src/workflow_engine/core/data.py:18]: contra and co?

Is something is missing here

**Comment by sunapi386** [comment]
This is a well-written PR description. It clearly explains the motivation and implementation. Here's my review:

- The C++ analogy effectively explains why automatic type casting would be valuable, making the "why" very understandable.
- Good Examples. The progression from simple (`StringValue("5")` → `IntegerValue(5)`) to complex (PDF files → text via context) examples illustrates the power of the system well. Although, passing context around reminds of Golang.
- Using Pydantic's RootModel as the foundation is good starting point.

## Areas for Review

### 1. **Missing Critical Information**
- **Breaking changes**: Will this affect existing code?
- **Performance implications**: Type casting at runtime could have overhead? What is that overhead? 
- **Error handling**: What happens when casts fail? How are errors propagated? What does the developer need to do?
- **Dependencies**: Are new dependencies being added?

### 2. **Implementation Details Need Clarification**
```python
SourceValue.add_cast(TargetValue, lambda S, T: lambda v, ctx: TargetValue(root=...))
```
This example is confusing:
- Why the nested lambda?
- What are S and T exactly, does it need to have exactly 2 lambdas? (The explanation is vague)
- When/where do developers call `add_cast`?

### 3. **Architecture Concerns**
- **Type safety**: How do you prevent runtime type errors?
- **Circular dependencies**: What prevents A→B→A casting loops?
- **Registration order**: Does the order of `add_cast` calls matter?
- **Context management**: How is context passed through the system?

### 4. **Missing Context**
- **Alternatives considered**: Why not use Python's existing type conversion patterns?
- **Integration plan**: How will this integrate with the existing node system?
- **Migration path**: How do existing nodes adopt this system?

## Suggestions

1. **Add a simple code example** showing the complete flow from registration to usage
2. **Include error handling examples** This being an open-source project, we need to hold ourselves to a high level of documentation so other developers can _hopefully_ use it with ease.
3. **Clarify the generic type handling** with a concrete example - in future PR perhaps.
4. **Add performance considerations** section
5. **Explain the context parameter** more thoroughly.. what counts as context?
6. Consider adding a brief section on limitations or known issues? Perhaps make use of the Discussions tab?

Overall, this is an ambitious and potentially very useful feature, but the PR description could benefit practical implementation details. Which I'm sure will be coming, as this is the 1st of many PRs.

**Review by sunapi386 — APPROVED** [review]
I had a somewhat difficult time reviewing this, feels like a little beyond my level of Python expertise. That being said, I prompted Sonnet 4 with the diff, and the description.

Take from the review below what you will.

### REVIEW

This is a substantial implementation with good test coverage. Here's my detailed review:

## Major Issues

### 1. **Dangerous Mutable Class Variables**
```python
_cast_cache: dict[OriginAndArgsRecursive, "Value"] = PrivateAttr(default_factory=dict)
```
This cache is **per-instance** but should probably be **per-class** or **global**. Currently, each Value instance has its own cache, which defeats the purpose of caching expensive operations across the system.

### 2. **Memory Leak Risk**
The `_cast_cache` grows indefinitely with no eviction policy. In a long-running system, this could consume significant memory.

### 3. **Thread Safety**
The caching mechanism and class-level `_resolved_casters` are not thread-safe. Multiple threads could corrupt the cache or cause race conditions during caster resolution.

### 4. **Fragile Generic Type Handling**
```python
if hasattr(cast_fn, "__code__") and cast_fn.__code__.co_argcount == 2:
```
This is brittle - it relies on introspection of function signatures rather than proper typing. Lambda functions, partial functions, or other callables might break this.

## Design Issues

### 1. **Inconsistent Error Handling**
```python
try:
    return cast_fn(from_args, to_args)
except Exception:
    logger.exception(f"Cannot instantiate cast function for type {t}")
    return None
```
Catching all exceptions and returning `None` hides important errors. The caller gets a generic "Cannot convert" message instead of the actual problem.

### 2. **Complex Registration API**
```python
StringValue.add_cast(
    IntegerValue, lambda S, T: lambda v, ctx: IntegerValue.model_validate_json(v.root)
)
```
The nested lambda pattern is confusing and error-prone. Consider a cleaner API:
```python
@StringValue.register_cast_to(IntegerValue)
def string_to_int(value: StringValue, context: Context) -> IntegerValue:
    return IntegerValue.model_validate_json(value.root)
```

### 3. **Questionable Inheritance of Casters**
```python
for parent in cls.__bases__:
    if issubclass(parent, Value):
        parent_casters = parent._get_casters()
```
This only looks at direct parents, not the full MRO. Also, inheriting casters might not always be desired behavior.

## Implementation Problems

### 1. **Incorrect Type Checking**
```python
def cast_sequence(Ss: Sequence[Type[Value]], Ts: Sequence[Type[V]]) -> Caster:
    (S,) = Ss  # This will crash if Ss has != 1 element
    (T,) = Ts
```
The tuple unpacking assumes exactly one element but doesn't validate this.

### 2. **Pydantic Metadata Access**
```python
info = t.__pydantic_generic_metadata__
```
This accesses private Pydantic internals that could change between versions, making the code brittle.

### 3. **Inefficient Caster Resolution**
The `_get_casters()` method rebuilds the entire caster dictionary on every call until cached, but the caching can be invalidated, causing repeated expensive operations.

## Test Issues

### 1. **Missing Edge Cases**
- No tests for circular casting dependencies
- No tests for malformed caster functions
- No tests for concurrent access
- No tests for memory usage with large caches

### 2. **Weak Error Testing**
```python
with pytest.raises(Exception):  # Too generic
```
Tests should verify specific exception types and messages.

## Positive Aspects

✅ **Comprehensive test coverage** for happy paths  
✅ **Good use of Pydantic** for serialization  
✅ **Clever use of RootModel** for transparent serialization  
✅ **Proper type annotations** throughout  
✅ **Immutable Value objects** prevent accidental mutations  

## Recommendations

### 1. **Fix the Caching Strategy**
```python
# Global cache with size limits
_CAST_CACHE: dict[tuple[Type, Type, int], Value] = {}
_CACHE_MAX_SIZE = 1000

def cast_to(self, t: Type[V], *, context: "Context") -> V:
    cache_key = (type(self), t, hash(self))
    if cache_key in _CAST_CACHE:
        return _CAST_CACHE[cache_key]
    
    # ... casting logic ...
    
    if len(_CAST_CACHE) >= _CACHE_MAX_SIZE:
        # Simple LRU eviction
        _CAST_CACHE.pop(next(iter(_CAST_CACHE)))
    
    _CAST_CACHE[cache_key] = result
    return result
```

### 2. **Improve Error Handling**
```python
class CastError(Exception):
    def __init__(self, from_type: Type, to_type: Type, reason: str):
        self.from_type = from_type
        self.to_type = to_type
        self.reason = reason
        super().__init__(f"Cannot cast {from_type} to {to_type}: {reason}")
```

### 3. **Simplify Generic Detection**
```python
from typing import Protocol

class SimpleCaster(Protocol):
    def __call__(self, value: Value, context: Context) -> Value: ...

class GenericCaster(Protocol):
    def __call__(self, from_args: Sequence[Type], to_args: Sequence[Type]) -> SimpleCaster: ...
```

### 4. **Add Validation**
```python
def add_cast(cls, t: Type[V], caster: GenericCaster | SimpleCaster):
    # Validate that the caster is callable
    if not callable(caster):
        raise TypeError("Caster must be callable")
    
    # Test the caster with dummy arguments to catch errors early
    try:
        if isinstance(caster, GenericCaster):
            test_fn = caster([], [])
            # Further validation...
    except Exception as e:
        raise ValueError(f"Invalid caster function: {e}")
```

## Overall Assessment

This is an ambitious and potentially useful feature, but it has several production-readiness issues around thread safety, memory management, and error handling. The core concept is sound, but the implementation needs hardening before it can be safely used in a production system.

The test coverage is good for basic functionality but needs expansion to cover edge cases and error conditions.

File: src/workflow_engine/core/value.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+    @classmethod
+    def add_cast(cls, t: Type[V], caster: GenericCaster):
+        """
+        Register a possible type cast from this class to the class T.
+        """
+        if cls._resolved_casters is not None:
+            raise RuntimeError(
+                f"Cannot add casters for {cls.__name__} after it has been used to cast values"
+            )
```
Comment (anthonytecsa) [inline src/workflow_engine/core/value.py:124]: I don't see a case where you'll need casting at runtime for workflows so this is good. But, this is not an obvious behaviour for users and should be well documented.

File: src/workflow_engine/core/data.py (hunk 1 above)
Comment (xujustinj) [reply src/workflow_engine/core/data.py:18]: just cleaning up some junk in the code, we typically always want to use `contra` and `co` suffixes on Python type variables for generic classes.

`covariant` means that `S is a T` implies `Foo[S] is a Foo[T]`

`contravariant` means that `S is a T` implies `Foo[T] is a Foo[S]`

Covariant is how we usually think about our container types: a `list[PDFFile]` is a `list[File]`

Contravariant is how we usually think about argument types of functions: a `Predicate[File]` is a `Predicate[PDFFile]` because any PDFfile you can pass into the latter can also be passed into the former; so the former serves at least the same functions

Comment (xujustinj) [reply src/workflow_engine/core/data.py:18]: To finish the mini infodump, Python actually gives special treatment to type variables that end in `_contra` and `_co`, which is why these guys are named weirdly -- it's extra reinforcement on top of the `covariant=True` and `contravariant=True` flags on those types

**Comment by xujustinj** [comment]
The suggestions for using protocols and decorators were :ok_hand: spot on. If it could just be more concise, AI reviews would be super awesome
